File: ts/src/verify-certificate/parse.ts
import { TheVeilCertificateError } from '../errors.js';
import type { VeilCertificate, VeilVerificationResult } from '../types.js';

// Structural validation — asserts the shape this arc reads. Does NOT
// validate enum literal membership beyond "string" here (the signable
// derivation layer throws `malformed` on unknown verdict literals, which
// is where that check belongs semantically). Returns the narrowed type
// on success or throws TheVeilCertificateError with reason 'malformed'.
export function parseCertificate(raw: unknown): VeilCertificate {
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
    throw new TheVeilCertificateError('Certificate is not a JSON object', {
      reason: 'malformed',
    });
  }
  const cert = raw as Partial<VeilCertificate>;
  const certId = typeof cert.certificate_id === 'string' ? cert.certificate_id : undefined;

  if (
    typeof cert.certificate_id !== 'string' ||
    typeof cert.request_id !== 'string' ||
    typeof cert.witness_key_id !== 'string' ||
    typeof cert.witness_signature !== 'string' ||
    typeof cert.issued_at !== 'string' ||
    typeof cert.protocol_version !== 'number' ||
    !Array.isArray(cert.claims) ||
    typeof cert.verification !== 'object' ||
    cert.verification === null
  ) {
    throw new TheVeilCertificateError('Certificate missing required fields', {
      reason: 'malformed',
      certificateId: certId,
    });
  }

  // C6 — verification.overall_verdict must be a string (protojson enum
  // form). The signable derivation step later maps this string to a Go
  // int enum; reject non-string values here rather than letting them
  // cascade into a less-legible invalid_signature.
  const verif = cert.verification as Partial<VeilVerificationResult>;
  if (typeof verif.overall_verdict !== 'string') {
    throw new TheVeilCertificateError(
      'verification.overall_verdict must be a string enum literal',
      { reason: 'malformed', certificateId: certId },
    );
  }

  return cert as VeilCertificate;
}

File: python/src/theveil/verify_certificate/parse.py
"""Structural parse of a raw Veil Certificate body."""

from __future__ import annotations

from typing import Any

from theveil.errors import TheVeilCertificateError
from theveil.types import VeilCertificate

__all__ = ["parse_certificate"]


def parse_certificate(raw: Any) -> VeilCertificate:
    """Structurally validate ``raw`` and return a :class:`VeilCertificate`.

    Asserts the shape this arc reads. Does NOT validate enum literal
    membership beyond "string" here — the signable derivation layer throws
    ``malformed`` on unknown verdict literals, which is where that check
    belongs semantically.

    Raises:
        TheVeilCertificateError: with ``reason="malformed"`` on bad shape
            or missing required fields.
    """

    if not isinstance(raw, dict):
        raise TheVeilCertificateError(
            "Certificate is not a JSON object",
            reason="malformed",
        )

    cert_id_raw = raw.get("certificate_id")
    cert_id = cert_id_raw if isinstance(cert_id_raw, str) else None

    required_str_fields = (
        "certificate_id",
        "request_id",
        "witness_key_id",
        "witness_signature",
        "issued_at",
    )
    for field_name in required_str_fields:
        if not isinstance(raw.get(field_name), str):
            raise TheVeilCertificateError(
                "Certificate missing required fields",
                reason="malformed",
                certificate_id=cert_id,
            )
    if not isinstance(raw.get("protocol_version"), int) or isinstance(
        raw.get("protocol_version"), bool
    ):
        # bool subclasses int in Python; exclude it so True/False never
        # passes as a protocol version.
        raise TheVeilCertificateError(
            "Certificate missing required fields",
            reason="malformed",
            certificate_id=cert_id,
        )
    if not isinstance(raw.get("claims"), list):
        raise TheVeilCertificateError(
            "Certificate missing required fields",
            reason="malformed",
            certificate_id=cert_id,
        )
    verification = raw.get("verification")
    if not isinstance(verification, dict):
        raise TheVeilCertificateError(
            "Certificate missing required fields",
            reason="malformed",
            certificate_id=cert_id,
        )
    if not isinstance(verification.get("overall_verdict"), str):
        raise TheVeilCertificateError(
            "verification.overall_verdict must be a string enum literal",
            reason="malformed",
            certificate_id=cert_id,
        )

    try:
        return VeilCertificate.model_validate(raw)
    except Exception as exc:  # pydantic.ValidationError etc.
        # Any Pydantic-level coercion failure on the declared fields
        # (e.g. claim with non-int nested field) surfaces as malformed.
        raise TheVeilCertificateError(
            f"Certificate validation failed: {exc}",
            reason="malformed",
            certificate_id=cert_id,
        ) from exc
